File: typescript-impl/src/infrastructure/validators/input_validator.ts
import { Gender } from '../enums/gender_choice';
import { AccountType } from '../enums/account_type';

export class InputValidator {

    static validateGender(genderInput: string): Gender | null {
        try {
            const gender = Gender[genderInput.toUpperCase() as keyof typeof Gender];
            return gender || null; // Return null if the input string doesn't match any enum value
        } catch (error) {
            return null;
        }
    }

    static validateDate(dateStr: string): Date | null {
        try {
            const date = new Date(dateStr);
            if (isNaN(date.getTime())) {
                return null;
            }
            return date;
        } catch (error) {
            return null;
        }
    }

    static validateEmail(email: string): string | null {
        const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
        if (emailRegex.test(email)) {
            return email;
        } else {
            return null;
        }
    }

    static validateAmount(amount: string): number | null {
        try {
            const parsedAmount = parseFloat(amount);
            if (isNaN(parsedAmount)) {
                return null;
            }
            return parsedAmount;
        } catch (error) {
            return null;
        }
    }

    static validateWithdrawAmount(amount: string, minimumWithdrawAmount: any, accountBalance: number): number | null {
        try {
            const parsedAmount = parseFloat(amount);
            const amountAfterWithdrawal = accountBalance - parsedAmount;
            if (isNaN(parsedAmount) || amountAfterWithdrawal < minimumWithdrawAmount) {
                return null;
            }
            return parsedAmount;
        } catch (error) {
            return null;
        }
    }

    static validateAccountType(accountTypeInput: string): AccountType | null {
        try {
            const accountType = AccountType[accountTypeInput.toUpperCase() as keyof typeof AccountType];
            return accountType || null;
        } catch (error) {
            return null;
        }
    }

    static validateAccountId(accountId: string): number | null {
        try {
            const parsedAccountId = parseInt(accountId);
            if (isNaN(parsedAccountId)) {
                return null;
            }
            return parsedAccountId;
        } catch (error) {
            return null;
        }
    }
}

File: insfrastructure/validators/input_validator.py
from datetime import datetime
import re
from insfrastructure.enums.gender_choice import Gender
from insfrastructure.enums.account_types import AccountType


class InputValidator:
    @staticmethod
    def validate_gender(gender_input):
        try:
            gender = Gender[gender_input.upper()]
            return gender
        except KeyError:
            return None

    @staticmethod
    def validate_date(date_str):
        try:
            date = datetime.strptime(date_str, "%Y-%m-%d")
            return date
        except ValueError:
            return None

    @staticmethod
    def validate_email(email):
        if re.match(r"[^@]+@[^@]+\.[^@]+", email):
            return email
        else:
            return None

    @staticmethod
    def validate_amount(amount):
        try:
            amount = float(amount)
            return amount
        except ValueError:
            return None

    @staticmethod
    def validate_withdraw_amount(amount, minimum_withdraw_amount, account_balance):
        try:
            amount = float(amount)
            amount_after_withdrawal = account_balance - amount
            if amount_after_withdrawal < minimum_withdraw_amount:
                return None
            return amount
        except ValueError:
            return None

    @staticmethod
    def validate_account_type(account_type_input):
        try:
            account_type = AccountType[account_type_input.upper()]
            return account_type
        except KeyError:
            return None

    @staticmethod
    def validate_account_id(account_id):
        try:
            account_id = int(account_id)
            return account_id
        except ValueError:
            return None
